File: src/silu_lut.cpp
// SuperSLM SwiGLU SiLU via a fixed-point sigmoid LUT (C10, S2.4).
//
// See include/superslm/silu_lut.h for the runtime contract and
// SuperSLM_S2.4_SiLU_LUT_Design §5 (index derivation) / §6 (interpolation) for the
// construction. Entirely integer, entirely division-free: the two roundings reuse the shipped
// RoundingDivideByPOT primitive (int64 to place the sub-node index when e+k+Q_idx < 0, int32 to
// interpolate), so C3's tie rule is not re-implemented here. The table is generated offline in
// double precision by the converter and read from the artifact's SIL1 section; the caller
// materializes it into the native-endian int32 array this lookup indexes.
#include "superslm/silu_lut.h"

#include "superslm/intmath.h"  // RoundingDivideByPOT (int32 + int64 overloads), the C3 tie rule

namespace superslm {

int32_t SiluSigmoidQ15(const int32_t* table, int8_t code, int64_t m, int e) {
	// §5 — the real-valued table position pos = (code·realscale + X)·K, in Q(kSiluLutQIdx) fixed
	// point. realscale = m·2^e and K = N/2X = 2^kSiluLutLog2K, so pos's scale contribution is
	// code·m·2^(e+k) — shift-and-add, never a general multiply-then-divide.
	const int64_t term = static_cast<int64_t>(code) * m;  // exact int64: |term| < 127·2^31 < 2^39
	const int shift = e + kSiluLutLog2K + kSiluLutQIdx;    // e + 5 + 12

	// Place the sub-node position. The branch is chosen by shift's sign: shift >= 0 (i.e.
	// e >= -(k+Q_idx) = -17) is the exact left-shift; shift < 0 rounds under C3 via the int64
	// RoundingDivideByPOT (the S2.4 prep overload). Two-sided caller-ensures precondition (§10
	// item 3), the same convention the other kernels use — a scale outside the swept range is a
	// site-level rejection (C29), not this leaf's guard: on the left branch `shift < 24` keeps
	// `term << shift` exact within int64 (|term| < 2^38, so |result| < 2^62); on the right branch
	// `-shift <= 63` stays in RoundingDivideByPOT's exponent domain. For every real calibrated
	// (m,e) the branch is shift < 0, with -shift measured in [15,19] over the full calibrated
	// corpus (§10 item 3 width sweep, 28 layers x 2 forwards); the only shift >= 0 inputs are
	// code == 0 (term 0) or code extremes that saturate below. The real compiled kernel overflows
	// the left branch at shift == 26 and hits the RoundingDivideByPOT domain limit at -shift == 64,
	// both far outside the corpus range (margin >= 2 shifts to the left edge, >= 44 to the right).
	int64_t pos_fixed = shift >= 0 ? (term << shift) : RoundingDivideByPOT(term, -shift);

	// + N/2 (in Q(kSiluLutQIdx)), then saturate to the table domain [0, N<<Q_idx] BEFORE the split.
	const int64_t domain_hi = static_cast<int64_t>(kSiluLutN) << kSiluLutQIdx;  // N << Q_idx
	pos_fixed += domain_hi / 2;                                                 // (N<<Q_idx)/2 (N even)
	if (pos_fixed < 0) pos_fixed = 0;
	else if (pos_fixed > domain_hi) pos_fixed = domain_hi;

	// §5 correction (§14 amendment 4): clamp the node index FIRST — the N-th node has no "hi"
	// neighbour past it — then derive frac against the CLAMPED index, so the two are one reference
	// frame. i0 in [0, N-1]; frac in [0, 2^Q_idx], reaching 2^Q_idx exactly at high saturation
	// (§6 resolves that to table[N]).
	int64_t i0_64 = pos_fixed >> kSiluLutQIdx;
	if (i0_64 > kSiluLutN - 1) i0_64 = kSiluLutN - 1;
	const int i0 = static_cast<int>(i0_64);
	const int64_t frac = pos_fixed - (static_cast<int64_t>(i0) << kSiluLutQIdx);

	// §6 — linear interpolation, exactly one C3 rounding. table[i0+1] always addresses a real
	// entry (i0 <= N-1, and the table has N+1 entries for exactly this reason). Under the
	// canonical table (ParseSigmoidLut's pinned-content check, S-HARDEN-1 F20/F22 — the ONLY
	// table this function is ever called with in the shipped path) |frac*diff| <= 2^12 * 2^15 =
	// 2^27, which fits int32 exactly as the design derives. `diff`/`product`/`delta` are widened
	// to int64_t as DEFENCE IN DEPTH beneath that content-pinning guarantee, not because the
	// canonical table needs it: the guarantee is load-bearing at the parser, not re-derived here,
	// but the two unvalidated-artifact-operand UB chain (F22) is the standing argument for never
	// letting a downstream arithmetic step be the last line of defense on its own.
	const int64_t lo = table[i0];
	const int64_t hi = table[i0 + 1];
	const int64_t diff = hi - lo;
	const int64_t product = frac * diff;
	const int64_t delta = RoundingDivideByPOT(product, kSiluLutQIdx);
	return static_cast<int32_t>(lo + delta);
}

}  // namespace superslm

File: tests/reference/superslm_spike/silu_lut.py
"""C10 -- the SwiGLU sigmoid fixed-point piecewise-linear LUT (S3.0, F-S3-1).

Ported per `Claude/Vitruvius/SuperSLM_S2.4_SiLU_LUT_Design-2026-07-19.md` Section 4 (the
table), Section 5 (the runtime index derivation), Section 6 (the runtime interpolation):
the pinned parameters (N=1024 nodes over x in [-16, 16], Q_idx=12 fractional bits, signed
Q15 table entries) and the two tie rules -- C14's round-half-even for the offline
table-generation step, C3's round-half-away-from-zero for both runtime roundings (the
index-placement shift and the interpolation).

`SuperSLM_S3a_WalkingSkeleton_Plan.md` Section 11 S3.0 (F-S3-1, Section 4.1) requires the
Python reference forward to compute this construction in place of the excluded
i-exp-sigmoid form (`intmath.i_exp_from_constants`-based). Both `pipeline.forward_dynamic`
and `dynamic_engine.forward_dynamic_vec` call `silu_sigmoid_q15` from this one module --
bit-equality between the two forwards' `mlp_act` sites depends on both calling the same
construction rather than two independently-written copies.

The offline table (`SIGMOID_Q15_TABLE`) is generated once, at import, in high-precision
decimal arithmetic -- the offline half of Section 3's offline/runtime split (the RoPE
`ROP1` table precedent this design follows). The runtime path below is integer-only,
division-free except for the two `intmath.rounding_divide_by_pot` roundings, and never
touches a float.
"""

from __future__ import annotations

from decimal import ROUND_HALF_EVEN, Decimal, localcontext
from typing import Sequence

from superslm_spike import intmath

__all__ = [
    "TABLE_N",
    "TABLE_X",
    "TABLE_Q_IDX",
    "TABLE_FRAC_BITS",
    "TABLE_K_SHIFT",
    "SIGMOID_Q15_TABLE",
    "silu_lut_index",
    "silu_lut_interpolate",
    "silu_sigmoid_q15",
]

# Section 4 -- PINNED for S2.4's initial ship.
TABLE_N = 1024
TABLE_X = 16
# Section 5 -- Q_idx = 12 fractional bits for the sub-node position.
TABLE_Q_IDX = 12
# Section 4 -- table entries are signed Q15 (SIGMOID_FRAC_BITS, pipeline.py:192).
TABLE_FRAC_BITS = 15

if TABLE_N % (2 * TABLE_X) != 0:
    raise AssertionError("Section 5 requires K = N / (2X) to be an exact integer")
_K = TABLE_N // (2 * TABLE_X)
if _K & (_K - 1) != 0:
    raise AssertionError("Section 5 requires K = N / (2X) to be a power of two")
# Section 5 -- "K = 2^k exactly ... the index scale is itself a power of two".
TABLE_K_SHIFT = _K.bit_length() - 1


def _round_half_even_to_int(value: Decimal) -> int:
    return int(value.quantize(Decimal(1), rounding=ROUND_HALF_EVEN))


def _build_sigmoid_q15_table(n: int = TABLE_N, x_bound: int = TABLE_X,
                              frac_bits: int = TABLE_FRAC_BITS) -> tuple[int, ...]:
    """Section 4's offline table, entry i (0..n) =
    round_half_even(sigmoid(-X + i*(2X/N)) * 2**frac_bits), at high (60-digit decimal)
    precision -- the one place this construction touches a non-integer type, exactly as
    RoPE's sin/cos table generation does (Section 3). Full-domain table, no
    mirror-and-negate (Section 4): n+1 entries, node n included as the closed top end.
    """
    with localcontext() as ctx:
        ctx.prec = 60
        step = Decimal(2 * x_bound) / Decimal(n)
        scale = Decimal(1 << frac_bits)
        one = Decimal(1)
        entries = []
        for i in range(n + 1):
            x = Decimal(-x_bound) + Decimal(i) * step
            sigmoid = one / (one + (-x).exp())
            entries.append(_round_half_even_to_int(sigmoid * scale))
    return tuple(entries)


SIGMOID_Q15_TABLE: tuple[int, ...] = _build_sigmoid_q15_table()


def silu_lut_index(code: int, m: int, e: int, n: int = TABLE_N,
                    q_idx: int = TABLE_Q_IDX, k_shift: int = TABLE_K_SHIFT
                    ) -> tuple[int, int]:
    """Section 5's index derivation: `(i0, frac)`, where `i0` is the CLAMPED node index --
    clamped first -- and `frac` is measured against that same clamped `i0` (the
    2026-07-19 frac/index-order correction). `frac` lands in the closed range
    `[0, 2**q_idx]`; it equals `2**q_idx` exactly (not 0) when `pos_fixed` saturates to
    the table's top end, which is what makes Section 6's saturation resolve to
    `table[n]` rather than `table[n-1]`.

    `code`: the int8 SwiGLU gate value in [-127, 127]. `(m, e)`: the gate's per-token
    carried scale, canonical form (`m` in [2**30, 2**31), integer `e`).
    """
    if not -127 <= code <= 127:
        raise ValueError(f"silu_lut_index: code out of int8 domain [-127,127]: {code}")
    if not (1 << 30) <= m < (1 << 31):
        raise ValueError(f"silu_lut_index: m out of canonical domain [2**30, 2**31): {m}")
    term = code * m
    shift = e + k_shift + q_idx
    if shift >= 0:
        pos_fixed = term << shift
    else:
        pos_fixed = intmath.rounding_divide_by_pot(term, -shift)
    pos_fixed += (n << q_idx) // 2
    pos_fixed = max(0, min(pos_fixed, n << q_idx))
    i0 = min(pos_fixed >> q_idx, n - 1)
    frac = pos_fixed - (i0 << q_idx)
    return i0, frac


def silu_lut_interpolate(table: Sequence[int], i0: int, frac: int,
                          q_idx: int = TABLE_Q_IDX) -> int:
    """Section 6's interpolation: `lo + RoundingDivideByPOT(frac * (hi - lo), q_idx)` --
    the C3 tie rule, the only rounding besides Section 5's own. `table` must have at
    least `i0 + 2` entries (the clamp in `silu_lut_index` guarantees `i0 <= n - 1`)."""
    lo, hi = table[i0], table[i0 + 1]
    diff = hi - lo
    product = frac * diff
    delta = intmath.rounding_divide_by_pot(product, q_idx)
    return lo + delta


def silu_sigmoid_q15(code: int, m: int, e: int, table: Sequence[int] = SIGMOID_Q15_TABLE,
                      n: int = TABLE_N, q_idx: int = TABLE_Q_IDX,
                      k_shift: int = TABLE_K_SHIFT) -> int:
    """C10's full runtime construction, composed: index derivation then interpolation
    (Section 5, Section 6) -- the reference forward's `mlp_act` conformance target
    (`SuperSLM_S3a_WalkingSkeleton_Plan.md` Section 11 S3.0)."""
    i0, frac = silu_lut_index(code, m, e, n=n, q_idx=q_idx, k_shift=k_shift)
    return silu_lut_interpolate(table, i0, frac, q_idx=q_idx)
